### PR TITLE
Eggsac carrier revival

### PR DESCRIPTION
**Pull request by Birdtalon**
[Forum Thread](https://forum.cm-ss13.com/t/eggsac-carrier-revival/4711)

# About the pull request

The concept of this PR is to find a middle ground between the current eggsac carrier and the pre #4459 eggsac carrier.

This PR will make the following changes. (From this point "normal weeds" can be substituted for "off hive weeds" Placing eggs on hive weeds is **unchanged**.)

- Eggsac carrier can once again place eggs on normal weeds.
- Eggsac carrier can only place 4 eggs at once on normal weeds.
- If the Eggsac places more than 4 eggs on normal weeds, their oldest placed egg will die. No hugger release.
- Eggs placed on normal weeds have a maximum lifetime of 5 MINUTES after which they will die. No hugger release.
- Eggs placed on normal weeds have a 1 MINUTE life whilst the carrier is further than 14 tiles away from them.
- If the carrier dies, all of their sustained eggs die.

# Explain why it's good for the game

Eggsac carrier at the moment is in a bit of a poor place and has gone from being very strong to quite poor. Considering the limitation of having to place only on hive weeds.
The majority of feedback I read against eggsac carrier was with the quantity of eggs able to be placed, as well as the locations they can be placed in, all across the map and with impunity.

This PR aims to address those concerns to make the eggsac both less infuriating to play against while still being satisfying to play as a frontline support or as a stealthy trapper.

Eggs can no longer be placed all over the map because of the 4 egg limit off weeds, so the carrier has to think where they want to impact the map. The carrier also has to stay within a reasonable distance to where they are impacting with their eggs which localises their impact to their immediate play area. The carrier also has to be more reactive to current events as they cannot place an egg which then becomes useful 30 minutes later.

Killing the carrier also has a small reward as in addition to removing a xeno from the game, the eggs they are sustaining are cleared. If a carrier is supporting the front and dies, the marines don't then have to clear X number of eggs AFTER the kill.

# Testing Photographs and Procedure

1. Spawn as egg carrier.
2. Plant egg on normal weeds.
3. Move 15+ tiles away.
4. Wait 1 minute
5. OR Wait 5 minutes within 14 tiles.
6. Kill the carrier.

The 5 minute lifetime of the eggs will also clear the eggs in the case that the carrier is admin deleted, or some other weird stuff happens which doesn't result in a death. This will also catch carriers de-evolving

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog

:cl:
add: Eggsac carrier can now place eggs on normal weeds to a maximum of 4 eggs.
add: Eggsac carrier eggs on normal weeds have an expiry date.
/:cl:

### CODE REVIEW

**Comment by Birdtalon** [comment]
Need to add: handling for eggs bursting for marines or being destroyed as I forgot this. Can't set to draft on mobile github.

**Comment by Birdtalon** [comment]
Updated with further handling I missed originally.

File: code/modules/cm_aliens/structures/egg.dm [hunk 1]
```diff
@@ -296,3 +296,68 @@
...
+SPECIAL EGG USED BY EGG CARRIER
+*/
+
+#define CARRIER_EGG_UNSUSTAINED_LIFE 1 MINUTES
+#define CARRIER_EGG_MAXIMUM_LIFE 5 MINUTES
+
+/obj/effect/alien/egg/carrier_egg
+	name = "fragile egg"
+	desc = "It looks like a weird, fragile egg."
+	var/owner = null
```
Comment (fira) [inline code/modules/cm_aliens/structures/egg.dm:310]: Explicit owner type, at last /xenomorph

File: code/modules/cm_aliens/structures/egg.dm [hunk 2]
```diff
@@ -296,3 +296,68 @@
...
+/obj/effect/alien/egg/carrier_egg/Initialize(mapload, hivenumber, planter = null)
+	. = ..()
+	last_refreshed = world.time
+	if(!planter)
+		//If we have no owner when created... this really shouldn't happen but start decaying the egg immediately.
+		start_unstoppable_decay()
+	else
+		//Die after maximum lifetime
+		life_timer = addtimer(CALLBACK(src, PROC_REF(remove_owner_and_decay)), CARRIER_EGG_MAXIMUM_LIFE, TIMER_STOPPABLE)
+		set_owner(planter)
```
Comment (fira) [inline code/modules/cm_aliens/structures/egg.dm:323]: If you keep a hard reference to owner around you need to be able to clear it up if they die. Eg. using COMSIG_PARENT_QDELETING

File: code/modules/cm_aliens/structures/egg.dm [hunk 3]
```diff
@@ -296,3 +296,68 @@
...
+	var/datum/behavior_delegate/carrier_eggsac/my_delegate = planter.behavior_delegate
+	my_delegate.eggs_sustained += src
+	owner = planter
+
+/obj/effect/alien/egg/carrier_egg/proc/remove_owner()
+	if(owner)
+		var/mob/living/carbon/xenomorph/carrier/my_owner = owner
+		var/datum/behavior_delegate/carrier_eggsac/my_delegate = my_owner.behavior_delegate
+		my_delegate.eggs_sustained -= src
+		owner = null
```
Comment (fira) [inline code/modules/cm_aliens/structures/egg.dm:341]: i think this would belong better on the behavior_delegate but it's just a detail

**Review by fira — CHANGES_REQUESTED** [review]

File: code/modules/cm_aliens/structures/egg.dm (hunk 1 above)
Comment (Birdtalon) [reply code/modules/cm_aliens/structures/egg.dm]: Done.

File: code/modules/cm_aliens/structures/egg.dm (hunk 3 above)
Comment (Birdtalon) [reply code/modules/cm_aliens/structures/egg.dm]: Moved to `behavior_delegate`.

File: code/modules/cm_aliens/structures/egg.dm (hunk 2 above)
Comment (Birdtalon) [reply code/modules/cm_aliens/structures/egg.dm:323]: Have added this signal, I also unregister it on `.../Burst()` as I've already setting the owner to null so it should just clean-up any cases where the carrier is deleted. Lmk if any issues with this.

File: code/modules/cm_aliens/structures/egg.dm [hunk 4]
```diff
@@ -296,3 +296,60 @@
...
+SPECIAL EGG USED BY EGG CARRIER
+*/
+
+#define CARRIER_EGG_UNSUSTAINED_LIFE 1 MINUTES
+#define CARRIER_EGG_MAXIMUM_LIFE 5 MINUTES
+
+/obj/effect/alien/egg/carrier_egg
+	name = "fragile egg"
+	desc = "It looks like a weird, fragile egg."
+	var/mob/living/carbon/xenomorph/carrier/owner = null
```
Comment (Zonespace27) [inline code/modules/cm_aliens/structures/egg.dm:310]: null this on destroy

File: code/modules/cm_aliens/structures/egg.dm [hunk 5]
```diff
@@ -296,3 +296,60 @@
...
+	if(!planter)
+		//If we have no owner when created... this really shouldn't happen but start decaying the egg immediately.
+		start_unstoppable_decay()
+	else
+		//Die after maximum lifetime
+		life_timer = addtimer(CALLBACK(src, PROC_REF(start_unstoppable_decay)), CARRIER_EGG_MAXIMUM_LIFE, TIMER_STOPPABLE)
+		set_owner(planter)
+
+/obj/effect/alien/egg/carrier_egg/Destroy()
+	. = ..()
```
Comment (Zonespace27) [inline code/modules/cm_aliens/structures/egg.dm:326]: call parent at the end of destroy

File: code/modules/cm_aliens/structures/egg.dm [hunk 6]
```diff
@@ -296,3 +296,60 @@
...
+	if(life_timer)
+		deltimer(life_timer)
+
+/obj/effect/alien/egg/carrier_egg/proc/set_owner(mob/living/carbon/xenomorph/carrier/planter)
```
Comment (Zonespace27) [inline code/modules/cm_aliens/structures/egg.dm:330]: code doc

File: code/modules/cm_aliens/structures/egg.dm [hunk 7]
```diff
@@ -296,3 +296,60 @@
...
+	var/datum/behavior_delegate/carrier_eggsac/my_delegate = planter.behavior_delegate
+	my_delegate.eggs_sustained += src
+	owner = planter
+	RegisterSignal(owner, COMSIG_PARENT_QDELETING, PROC_REF(cleanup_owner))
```
Comment (Zonespace27) [inline code/modules/cm_aliens/structures/egg.dm:334]: you don't need to do this, you can just do the destroy bit i mentioned above

File: code/modules/cm_aliens/structures/egg.dm [hunk 8]
```diff
@@ -296,3 +296,60 @@
...
+	addtimer(CALLBACK(src, PROC_REF(Burst), TRUE), 10 SECONDS)
+	if(life_timer)
+		deltimer(life_timer)
+
+/obj/effect/alien/egg/carrier_egg/Burst(kill, instant_trigger, mob/living/carbon/xenomorph/X, is_hugger_player_controlled)
+	. = ..()
+	UnregisterSignal(owner, COMSIG_PARENT_QDELETING)
+	owner = null
+
+/obj/effect/alien/egg/carrier_egg/proc/cleanup_owner()
```
Comment (Zonespace27) [inline code/modules/cm_aliens/structures/egg.dm:352]: code doc

File: code/modules/mob/living/carbon/xenomorph/mutators/strains/carrier/eggsac.dm [hunk 9]
```diff
@@ -40,8 +43,44 @@
...
 	return TRUE
 
+/datum/behavior_delegate/carrier_eggsac
+	name = "Eggsac Carrier Behavior Delegate"
+	var/list/eggs_sustained = list()
+	var/egg_sustain_cap = EGGSAC_OFF_WEED_EGGCAP
+	var/sustain_distance = 14
+
+/datum/behavior_delegate/carrier_eggsac/append_to_stat()
+	. = "Eggs sustained: [length(eggs_sustained)] / [egg_sustain_cap]"
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/xenomorph/mutators/strains/carrier/eggsac.dm:56]: are you sure it's = instead of +=?

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Comment by Zonespace27** [comment]
Willing to give this a shot.

File: code/modules/mob/living/carbon/xenomorph/mutators/strains/carrier/eggsac.dm (hunk 9 above)
Comment (Birdtalon) [reply code/modules/mob/living/carbon/xenomorph/mutators/strains/carrier/eggsac.dm:56]: I don't think it actually matters in this case. But comparing it to the same functionality on other behavior delegates I should be consistent with them, also returning it as a list.

https://github.com/Birdtalon/cmss13/blob/c7e8ff476a2fbae3654794173d38f941a2fcc775/code/modules/mob/living/carbon/xenomorph/XenoProcs.dm#L88-L90

**Comment by Birdtalon** [comment]
Think I've addressed all your comments.

- Remove the signal handler
- null owner on destroy
- Added code doc where requested
- Returning list and `+=` on `append_to_stat()`

File: code/modules/cm_aliens/structures/egg.dm [hunk 10]
```diff
@@ -296,3 +296,55 @@
...
+
+#define CARRIER_EGG_UNSUSTAINED_LIFE 1 MINUTES
+#define CARRIER_EGG_MAXIMUM_LIFE 5 MINUTES
+
+/obj/effect/alien/egg/carrier_egg
+	name = "fragile egg"
+	desc = "It looks like a weird, fragile egg."
+	var/mob/living/carbon/xenomorph/carrier/owner = null
+	var/last_refreshed = null
+	var/life_timer = null
```
Comment (Zonespace27) [inline code/modules/cm_aliens/structures/egg.dm:312]: code doc these

File: code/modules/cm_aliens/structures/egg.dm [hunk 11]
```diff
@@ -296,3 +296,55 @@
...
+	else
+		//Die after maximum lifetime
+		life_timer = addtimer(CALLBACK(src, PROC_REF(start_unstoppable_decay)), CARRIER_EGG_MAXIMUM_LIFE, TIMER_STOPPABLE)
+		set_owner(planter)
+
+/obj/effect/alien/egg/carrier_egg/Destroy()
+	if(life_timer)
+		deltimer(life_timer)
+	owner = null
+	. = ..()
```
Comment (Zonespace27) [inline code/modules/cm_aliens/structures/egg.dm:329]:
```suggestion
	return ..()
```

File: code/modules/mob/living/carbon/xenomorph/mutators/strains/carrier/eggsac.dm [hunk 12]
```diff
@@ -40,8 +43,45 @@
...
 	carrier.eggs_max = 12
 	carrier.egg_planting_range = 2
+	apply_behavior_holder(carrier)
 	return TRUE
 
+/datum/behavior_delegate/carrier_eggsac
+	name = "Eggsac Carrier Behavior Delegate"
+	var/list/eggs_sustained = list()
+	var/egg_sustain_cap = EGGSAC_OFF_WEED_EGGCAP
+	var/sustain_distance = 14
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/xenomorph/mutators/strains/carrier/eggsac.dm:53]: code doc

**Review by Zonespace27 — CHANGES_REQUESTED** [review]
a few nits left

**Comment by Birdtalon** [comment]
Addressed your comments.

**Comment by HeresKozmos** [comment]
nice!

**Pull request merged**